docs/melange-book/counter.md

```diff
@@ -0,0 +1,239 @@
+
+<!--#prelude#
+let (counter, setCounter) = React.useState(() => 0);
+let _ =
+-->
+```reasonml
+{counter |> Int.to_string |> React.string}
+```
+
+This uses the [pipe last operator](../communicate-with-javascript#pipe-last),
+which is useful for chaining function calls.
+
+Let's add a bit of styling to the root element of `Counter`:
+
+<!--#prelude#
+let (counter, setCounter) = React.useState(() => 0);
+let _ =
+-->
+```reasonml
+<div
+  style={ReactDOMStyle.make(
+    ~padding="1em",
+    ~display="flex",
+    ~gridGap="1em",
+    (),
+  )}>
+  <button onClick={_evt => setCounter(v => v - 1)}>
+    {React.string("-")}
+  </button>
+  <span> {counter |> Int.to_string |> React.string} </span>
+  <button onClick={_evt => setCounter(v => v + 1)}>
+    {React.string("+")}
+  </button>
+</div>;
+```
+
+Unlike in React, the `style` prop in ReasonReact doesn't take a generic object,
+instead it takes an object of type `ReactDOMStyle.t` that is created by calling
+`ReactDOMStyle.make`. This isn't a sustainable way to style our app---in future
+sections we will see how to style using CSS classes.
+
+Congratulations! You've created your first ReasonReact app and component. In
+future chapters we'll create more complex and interesting components.
+
+## Exercises
+
+<b>1.</b> There aren't any runtime errors in our app right now. But what happens if you
+   try to remove the `| None` branch of the `switch (node)` expression?
+
+<b>2.</b> What happens if you rename the `_evt` variable inside the button callback to
+   `evt`?
+
+<b>3.</b> Comment out the `[@react.component]` attribute in `Counter.re`. What happens?
+
+## Overview
+
+What we covered in this section:
+
+- How to create and run a basic ReasonReact app
+- ReasonReact components are also modules
+- OCaml has an `option` type whose value can be either `None` or `Some(_)`
+- The pipe last operator is an alternate way to invoke functions that enables
+  easy chaining of function calls
+- The `style` prop doesn't take generic objects
+
+## Solutions
+
+<b>1.</b> Removing the `| None` branch will result in a compilation error:
+
+```
+Error (warning 8 [partial-match]): this pattern-matching is not exhaustive.
+Here is an example of a case that is not matched:
+None
+```
+
+Basically, the compiler is telling you to handle the `None` case if you want
+to ship your app. This is part of what makes OCaml such a type-safe language.
+
+<b>2.</b> Renaming `_evt` to `evt` results in a compilation error:
+
+```
+Error (warning 27 [unused-var-strict]): unused variable evt.
+```
+
+OCaml wants you to use all the variables you declare, unless they begin with
+`_` (underscore).
+
+<b>3.</b> Commenting out `[@react.component]` in `Counter.re` will trigger a
+compilation error in `Index.re`, at the place where `Counter` component is used:
+
+```
+File "src/Index.re", line 3, characters 19-27:
+3 |   let make = () => <Counter />;
+                      ^^^^^^^^
+Error: Unbound value Counter.makeProps
```

Comment:
I think this error/solution might leave readers scratching their heads. What is `makeProps` and where it's coming from? But maybe it makes sense once the PPX chapter is ready, idk. 🤔